### DISCUSSION TITLE
Implementing an algebraic expectations operator

### DISCUSSION THREAD

**Discussion by jessegrabowski** [sympy/sympy#26896] [Q&A] [answered]
Hi all,

I'm using Sympy to work with and manipulate time series models. I've been using an extended Symbol (`TimeAwareSymbol`) to reason about lag/lead operators in deterministic systems -- [see here](https://gist.github.com/jessegrabowski/2e0a81ad0e52b9058755bb648e229e40) for an example of what I'm doing.

I'm interested in implementing an expectation operator that would help take this to the next level, by thinking about random variables inside these models. I know there is `sympy.stats.E`, but this actually evaluates random variables. I would like something that is purely symbolic and can play nice with simplification operations (e.g. it is known to be linear function, so sympy will gather up $E[X] + E[aY] = E[X + aY]$ but not $E[X] \cdot E[Y]$).

I guess I'm asking to be pointed towards resources/docs for extending sympy to add new operators. Thanks for any help!

**Comment by oscarbenjamin** [comment, marked as answer]
I think what is needed is not a new operator `E` but rather the ability to create a symbol representing a random variable with no specified probability distribution. There is a way to do this although more could be done to make it useful and it could perhaps be better documented:
```python
In [5]: from sympy.stats.rv import RandomSymbol

In [6]: from sympy.stats import *

In [7]: X = RandomSymbol('X')

In [8]: Y = RandomSymbol('Y')

In [9]: E(2*X + Y)
Out[9]: E[2⋅X] + E[Y]

In [10]: E(X*Y)
Out[10]: E[X⋅Y]

In [11]: E(X)*E(Y)
Out[11]: E[X]⋅E[Y]
```

**Comment by jessegrabowski** [comment]
Oh even better! I'll give this a try, but it looks like exactly what I want.
